Adds alternate behaviours to addtimer, making it even more useful than before.

nullbear opened the pull request:
Also updates camera.dm to use the timer.

DNM i broke something.

WHAT IS THIS?

this adds more functionality to proc/addtimer, allowing for different behaviours.
to be used in place of spawn(x) for increased code reuseability, quality, and stability.

addtimer(src, "procname", timetowait, unique, BEHAVIOUR)

When using spawn, it can typically be replaced with this.

**old:**

```
stuffity stuff
spawn(5)
        more stuff
        even more stuff
other stuff
```

**new:**

```
stuffity stuff
addtimer(src, "dostuff", 5)
otherstuff

proc/dostuff()
    more stuff
    even more stuff
```

the fourth argument can be set to ANYTHING in order to use the behaviours, which allow you to extend or shorten timers, or alter the arguments of the timer before it finishes. 

BEHAVIOUR argument:

the behaviour arguments are only compatible with the defined behaviours. This is used to control matched unique timers.

**TIMER_DEFAULT** will never overwrite an existing timer. this is the default behaviour.
**TIMER_WAIT** will overwrite timer only. Useful for extending a timer without altering it.
**TIMER_DATA** will overwrite arguments only. Useful for altering without extending it.
**TIMER_FULL** will overwrite arguments and timer. Useful for completely altering it.

_I HAVE NO IDEA HOW TO TEST THIS or IF IT WORKS. was guided by @MrStonedOne  and @RemieRichards_

nullbear committed:
Allows replacing unique timer with newest, instead of oldest only.

Also updates camera.dm to use the timer.

I HAVE NO IDEA HOW TO TEST THIS or IF IT WORKS. was guided by mso and
remierichards

RemieRichards commented on code/controllers/subsystem/timer.dm:
```diff
@@ -48,9 +53,11 @@ var/datum/subsystem/timer/SStimer
 /datum/timedevent/Destroy()
 	SStimer.processing -= src
 	SStimer.hashes -= hash
+	SStimer.unique[hash] = null
```
this line isn't needed, because of the line below.

nullbear committed:
Update timer.dm

coiax commented:
I have no idea what this PR does or is for.

nullbear commented:
It adds to the timer subsystem and addtimer() proc. Its really just improving code quality and reuseability.

Menshin commented on code/controllers/subsystem/timer.dm:
```diff
@@ -68,9 +74,17 @@ var/datum/subsystem/timer/SStimer
+		if(TIMER_NEWEST)
+			var/datum/timedevent/old = SStimer.unique[event.hash]
+			qdel(old)
+			SStimer.processing -= old // In case qdel doesnt get to it fast enough to remove it from the processing list.
```
Makes more sense to remove reference before ```qdel```'ing.

Menshin commented on code/game/machinery/camera/camera.dm:
```diff
@@ -79,33 +79,29 @@
 		if(prob(150/severity))
 			icon_state = "[initial(icon_state)]emp"
-			var/list/previous_network = network
+			networkbackup = network
```
Are you sure this actually copy the list ?

RemieRichards replied:
it doesn't copy it no, it simply references the same list object.

Menshin replied:
Then i guess there's no problem with ```network = list()``` next line...

RemieRichards replied:
Honk.

nullbear replied:
I'm assuming it does, since this is what was used before.

Menshin commented on code/game/machinery/camera/camera.dm:
```diff
@@ -79,33 +79,29 @@
 
+/obj/machinery/camera/proc/emp_reset()
+	triggerCameraAlarm()
+	network = networkbackup
```
Same question here.

RemieRichards replied:
Same answer above.

nullbear committed:
Added TIMER_SHORTEST and TIMER_LONGEST options to timer subsystem.

Intended uses:

Use TIMER_DEFAULT if overlapping timers are allowed.

Use TIMER_OLDEST to ignore newly created timers completely. Preventing them from being shortened or extended.
Use TIMER_NEWEST to be able to overwrite the timer with new timers, allowing them to be shortened or extended.
Use TIMER_SHORTEST to use the timer that will fire first only, preventing timers from being extended, but allowing them to be shortened.
Use TIMER_LONGEST to use the timer that will fire last only, preventing timers from being shortened, but allowing them to be extended.

Menshin commented on code/controllers/subsystem/timer.dm:
```diff
@@ -68,9 +76,37 @@ var/datum/subsystem/timer/SStimer
+			var/datum/timedevent/old = SStimer.unique[event.hash]
+			if(old)
+				qdel(old)
+			SStimer.processing -= old // In case qdel doesnt get to it fast enough to remove it from the processing list.
```
Makes more sense for it to happen into the ```if(old)```.
Same below.

nullbear replied:
I'm not really sure what you mean.

Menshin replied:
```if(old)``` fails when ```old``` is ```null```.
Then ```SStimer.processing -= null``` doesn't really do much.

nullbear replied:
but it wont cause an error, either. I was concerned that trying to qdel(null) might be an issue. And for the later ones, trying to access null.vars might be an issue.

removing null from a list isnt an issue. However removing it should probably be done BEFORE qdelling it.

nullbear committed:
Moves SStimer.processing before Qdel

nullbear commented:
Alright i've fixed that for you, menshin

nullbear committed:
Update camera.dm

MrStonedOne commented:
TIMER_SHORTEST and TIMER_LONGEST are pointless.

The timetowait arg is part of the hash, so only things with the same timetowait will ever match, meaning that SHORTEST/LONGEST are the same as  OLDEST/NEWEST

MrStonedOne commented:
One idea would be to make behavior be a simi bitflag so that this could be changed to make it so timetowait could be not part of the hash or part of the hash, depending,

nullbear commented:
would it be better to not have timetowait as part of the hash period? Everything would work then, but stuff like addtimer(x, x, 40) and addtimer(x, x, 10) would conflict under TIMER_OLDEST, whereas they don't currently, despite all other args being the same.

MrStonedOne commented:
it might be,

nullbear committed:
Update timer.dm

nullbear committed:
unique timer is reimplemented.

Set unique to ANYTHING but the defined behaviours, and the timer will only ever allow for one to exist.

nullbear committed:
Update timer.dm

nullbear committed:
Update timer.dm

nullbear committed:
Update camera.dm

nullbear commented:
@MrStonedOne okay i think that makes more sense now. How about you?
